### ISSUE TITLE
Setup Branch Protection for `main`

### ISSUE THREAD

**Issue by sebastian-sommerfeld-io** [sommerfeld-io/source2adoc#2]
As the owner and maintainer of the `source2adoc` project, which resides within the "sommerfeld-io" organization on GitHub,
I want to allow contributions from the open-source community
So that the project can benefit from a wider range of perspectives and improvements.

### Acceptance Criteria
- As a member of the "sommerfeld-io" organization ...
    - .... I should be able to read the repository.
    - .... I should be able to push changes directly to the repository and to create branches.
    - .... I should be able to create tags.
- As a non-member of the "sommerfeld-io" organization ...
    - .... I should be able to read the repository.
    - .... I should not be able to push changes directly to the repository or to create branches.
    - .... I should not be able to create tags.
- As a contributor ....
    - .... I should be able to fork the repository.
    - .... I should be able to introduce changes through pull requests from my fork.
- Pull requests should undergo review by repository maintainers before merging.
    - I should be able to review my own pull requests

#### Additional Requirements

- Branch Protection
    - Configure branch protection rules for the main branch to prevent direct pushes from non-members.
    - Specify that only members of the "sommerfeld-io" organization have write access to the main branch.
    - Require pull request reviews from members of the organization before merging.
- Collaborator Permissions
    - Ensure that only members of the "sommerfeld-io" organization have write access to the repository.
    - Remove external collaborators who shouldn't have write access.
- Code Review Policies
    - Enforce code review policies for pull requests.
    - Require a certain number of approving reviews from members of the organization before a pull request can be merged.

### Additional Todo
- [ ] Update `contribute.adoc` (draft see below)
- [ ] Implement for https://github.com/sommerfeld-io/source2adoc-website/issues/4 as well
    - [ ] Check if this can be done through Github CLI
- [ ] Document in https://github.com/sommerfeld-io/website/blob/main/docs/modules/howto/pages/setup-new-repo.adoc
- [ ] Fork repo with second account and test how the pipelines and secrets and pushing and everything behave
- [ ] Roll out to all repos of the sommerfeld-io organization
- [ ] Check if my personal projects need similar settings

By implementing these measures, we can maintain control over who can directly modify the repository while still allowing contributions from the open-source community through the forking and pull request workflow.

**Comment by sebastian-sommerfeld-io** [comment]
```
= How to Contribute to "source2adoc" Project

== Introduction
Thank you for your interest in contributing to the "source2adoc" project! This guide outlines the steps to contribute effectively while adhering to the project's contribution guidelines and restrictions.

== Prerequisites
Before you begin, ensure that you have the following:

- A GitHub account.
- Familiarity with Git and GitHub.
- Basic understanding of AsciiDoc format

== Contribution Workflow
. **Fork the Repository**:
.. Visit the `source2adoc` repository on GitHub: https://github.com/sommerfeld-io/source2adoc
.. Click on the "Fork" button in the top-right corner to create a fork of the repository under your GitHub account.
. **Clone Your Fork**:
.. Clone your forked repository to your local machine using the following command: `git clone https://github.com/your-username/source2adoc.git`
. **Make Changes**:
.. Create a new branch for your changes: `git checkout -b your-branch-name`
.. Make your desired changes to the code and/or documentation using your preferred text editor or IDE.
. **Commit Your Changes**:
.. Once you've made your changes, stage them for commit: `git add .`
.. Commit your changes with a descriptive commit message: `git commit -m "Brief description of your changes"`
. **Push Changes to Your Fork**:
.. Push your changes to your forked repository on GitHub: `git push origin your-branch-name`
. **Submit a Pull Request**:
.. Visit your forked repository on GitHub.
.. Click on the "Compare & pull request" button next to your branch.
.. Provide a descriptive title and detailed description for your pull request.
.. Submit the pull request for review.
. **Wait for Review**:
.. Your pull request will undergo review by the project maintainers.
.. Collaborate in discussions about changes or feedback during the review process.
. **Merge Pull Request**:
.. Once your pull request is approved, it will be merged into the main branch of the "source2adoc" repository.

== Contribution Guidelines

. Ensure your code adheres to the project's coding standards and conventions.
. Provide clear and concise commit messages.
. Test your changes thoroughly before submitting a pull request - ideally through Test Driven Development
. Be open to feedback and constructive criticism from project maintainers and contributors.

== Restrictions

. Non-members of the "sommerfeld-io" organization are not allowed to push directly to the repository or create branches.
. Tag creation is restricted to members of the organization.

== Get Help
If you need assistance or have any questions, feel free to reach out to the project maintainers via GitHub issues or email.

Thank you for contributing to the "source2adoc" project!
```

**Comment by sebastian-sommerfeld-io** [comment]
# ChatGPT proposes this solution

To achieve the requirements outlined in the user story for your "source2adoc" project on GitHub, follow these steps:

1. **Set Repository Permissions:**
   - Navigate to the repository settings on GitHub (https://github.com/sommerfeld-io/source2adoc/settings).
   - Under the "Manage access" section, ensure that only members of the "sommerfeld-io" organization have write access to the repository. Remove any external collaborators who shouldn't have write access.

2. **Configure Branch Protection:**
   - In the repository settings, go to the "Branches" section.
   - Add a branch protection rule for the main branch (e.g., `main` or `master`).
   - Check the box to "Require pull request reviews before merging" and specify that only members of the organization can approve pull requests.
   - Optionally, enforce additional restrictions such as requiring status checks to pass before merging.

3. **Enable Forking Workflow:**
   - By default, GitHub allows forking of public repositories.
   - Contributors from outside the organization can fork your repository by clicking the "Fork" button on the repository's page.
   - They can make changes in their forked repository and submit pull requests to the main repository for review.

4. **Restrict Tag Creation:**
   - Tag creation permissions are typically tied to write access to the repository.
   - Since only members of the organization have write access, external contributors won't be able to create tags directly.
   - You can monitor and manage tags manually as needed.

5. **Code Review and Merging:**
   - When a contributor submits a pull request, members of the organization can review the changes.
   - Once the pull request is approved, a member can merge it into the main branch.

6. **Automate Checks and Security Scanning:**
   - Set up CI/CD pipelines and code quality checks to run on pull requests automatically.
   - Integrate security scanning tools to check for vulnerabilities in contributed code.

7. **Documentation and Communication:**
   - Clearly communicate the contribution guidelines and workflow to potential contributors.
   - Provide guidance on how to submit high-quality pull requests.
   - Be responsive to issues and pull requests submitted by contributors, providing feedback and assistance as needed.

By following these steps, you can create a controlled contribution environment for your "source2adoc" project, allowing contributions from the open-source community while maintaining control over who can directly modify the repository.
